Problem: multiple SNS pools are not supported

vvv opened the pull request:
### 🧑‍🔬 Note to QA engineers and Hare maintainers

This patch should not break LDR1 setups (they use single SNS pool).  Please run the usual set of LDR1 tests (preferably on HW setup) and if they pass — merge the patch.  It's OK to postpone testing of multipool configurations.  **The longer this branch remains unmerged, the more it rots.**

---

Context:
- _Pools_ refer to non-overlapping sets of disks.
- _Profile_ is a reference to one or more pools.
- When Motr client is started, it receives profile fid from the command line.  Motr client will only be able to use those pools, which are referred to by its profile.

So far `cfgen` was only able to generate Motr configuration with a single SNS pool and a single profile.  This patch adds multipools support.

Solution:
* Extend CDF format so that multiple SNS pools and profiles can be specified:
  - add optional `disk_refs` field to pool descriptor.  `disk_refs` assign disks to SNS pools;
  - add optional `profiles` section.
* Update `hctl status` to show fids and names of profiles and SNS pools.
* Modify Consul KV schema:
  - revise `m0conf/profiles/` entries;
  - add `m0conf/pools/` entries;
  - update 4/KV RFC correspondingly.

![global config update](https://user-images.githubusercontent.com/17462/97111000-f002c500-16e4-11eb-9c1c-a4c457b3252a.jpeg)

Some uses of multiple pools:
* Distribute pools between groups of users.  When users work with different pools, they are less likely to overwrite each other's data.
* Use different striping patterns: 8+2 for one pool, 16+2 for another, etc.
* Hardware addition --- new disks form new pool.

---

Jira: [EOS-13951](https://jts.seagate.com/browse/EOS-13951)

vvv commented:
`singlenode-multipools.yaml` file that I used for testing multipools:

```yaml
# Cluster Description File (CDF).
# See `cfgen --help-schema` for the format description.

nodes:
  - hostname: localhost     # [user@]hostname
    data_iface: eth1        # name of data network interface
    #data_iface_type: o2ib  # type of network interface (optional);
                            # supported values: "tcp" (default), "o2ib"
    m0_servers:
      - runs_confd: true
        io_disks:
          data: []
      - io_disks:
          #meta_data: /path/to/meta-data/drive
          data:
            - /dev/loop0
            - /dev/loop1
            - /dev/loop2
            - /dev/loop3
            - /dev/loop4
            - /dev/loop5
            - /dev/loop6
            - /dev/loop7
            - /dev/loop8
            - /dev/loop9
    m0_clients:
      s3: 0         # number of S3 servers to start
      other: 2      # max quantity of other Motr clients this host may have
pools:
  - name: even
    #type: sns  # optional; supported values: "sns" (default), "dix", "md"
    disk_refs:
      - { path: /dev/loop0, node: localhost }
      - path: /dev/loop2
      - path: /dev/loop4
      - path: /dev/loop6
      - path: /dev/loop8
    data_units: 1
    parity_units: 0
    #allowed_failures: { site: 0, rack: 0, encl: 0, ctrl: 0, disk: 0 }
  - name: odd
    disk_refs:
      - path: /dev/loop1
      - path: /dev/loop3
      - path: /dev/loop5
      - path: /dev/loop7
      - path: /dev/loop9
    data_units: 1
    parity_units: 0
    #allowed_failures: { site: 0, rack: 0, encl: 0, ctrl: 0, disk: 0 }
profiles:
  - name: first-pool
    pools: [ even ]
  - name: both-pools
    pools: [ even, odd ]
```

See also `cfgen/examples/multipools.yaml`.

vvv commented:
cc @andriytk, @nezzzu

vvv commented:
cc @madhavemuri, @just-now

andriytk reviewed (approved):
Let's merge it to main and test it...

andriytk commented:
Singlenode works fine:

```
[vagrant@cmu hare]$ hctl bootstrap --mkfs ~/singlenode.yaml 
2020-10-22 15:20:34: Generating cluster configuration... OK
2020-10-22 15:20:35: Starting Consul server on this node............... OK
2020-10-22 15:20:49: Importing configuration into the KV store... OK
2020-10-22 15:20:49: Starting Consul on other nodes... OK
2020-10-22 15:20:49: Updating Consul configuraton from the KV store... OK
2020-10-22 15:20:50: Waiting for the RC Leader to get elected..... OK
2020-10-22 15:20:52: Starting Motr (phase1, mkfs)... OK
2020-10-22 15:21:01: Starting Motr (phase1, m0d)... OK
2020-10-22 15:21:04: Starting Motr (phase2, mkfs)... OK
2020-10-22 15:21:16: Starting Motr (phase2, m0d)... OK
2020-10-22 15:21:19: Checking health of services... OK
```

andriytk commented:
Started the cluster on two pools also:

```
[vagrant@cmu ~]$ hctl bootstrap --mkfs ~/singlenode-mp.yaml 
2020-10-22 15:27:33: Generating cluster configuration... OK
2020-10-22 15:27:35: Starting Consul server on this node............. OK
2020-10-22 15:27:46: Importing configuration into the KV store... OK
2020-10-22 15:27:46: Starting Consul on other nodes... OK
2020-10-22 15:27:46: Updating Consul configuraton from the KV store... OK
2020-10-22 15:27:47: Waiting for the RC Leader to get elected....... OK
2020-10-22 15:27:51: Starting Motr (phase1, mkfs)... OK
2020-10-22 15:27:58: Starting Motr (phase1, m0d)... OK
2020-10-22 15:28:00: Starting Motr (phase2, mkfs)... OK
2020-10-22 15:28:10: Starting Motr (phase2, m0d)... OK
2020-10-22 15:28:13: Checking health of services... OK
[vagrant@cmu ~]$ hctl status
Data pools:
    # fid name
    0x6f00000000000001:0x2e 'even'
    0x6f00000000000001:0x39 'odd'
Profiles:
    # fid name: pool(s)
    0x7000000000000001:0x54 'first-pool': 'even'
    0x7000000000000001:0x55 'both-pools': 'even' 'odd'
Services:
    localhost  (RC)
    [started]  hax        0x7200000000000001:0x6   192.168.180.145@tcp:12345:1:1
    [started]  confd      0x7200000000000001:0x9   192.168.180.145@tcp:12345:2:1
    [started]  ioservice  0x7200000000000001:0xc   192.168.180.145@tcp:12345:2:2
    [unknown]  m0_client  0x7200000000000001:0x28  192.168.180.145@tcp:12345:4:1
    [unknown]  m0_client  0x7200000000000001:0x2b  192.168.180.145@tcp:12345:4:2
[vagrant@cmu ~]$ cat ~/singlenode-mp.yaml 
# Cluster Description File (CDF).
# See `cfgen --help-schema` for the format description.

nodes:
  - hostname: localhost     # [user@]hostname
    data_iface: eth0        # name of data network interface
    #data_iface_type: tcp  # type of network interface (optional);
                            # supported values: "tcp" (default), "o2ib"
    m0_servers:
      - runs_confd: true
        io_disks:
          data: []
      - io_disks:
          #meta_data: /path/to/meta-data/drive
          data:
            - /dev/loop0
            - /dev/loop1
            - /dev/loop2
            - /dev/loop3
            - /dev/loop4
            - /dev/loop5
            - /dev/loop6
            - /dev/loop7
            - /dev/loop8
            - /dev/loop9
    m0_clients:
      s3: 0         # number of S3 servers to start
      other: 2      # max quantity of other Motr clients this host may have
pools:
  - name: even
    #type: sns  # optional; supported values: "sns" (default), "dix", "md"
    disk_refs:
      - { path: /dev/loop0, node: localhost }
      - path: /dev/loop2
      - path: /dev/loop4
      - path: /dev/loop6
      - path: /dev/loop8
    data_units: 1
    parity_units: 0
    #allowed_failures: { site: 0, rack: 0, encl: 0, ctrl: 0, disk: 0 }
  - name: odd
    disk_refs:
      - path: /dev/loop1
      - path: /dev/loop3
      - path: /dev/loop5
      - path: /dev/loop7
      - path: /dev/loop9
    data_units: 1
    parity_units: 0
    #allowed_failures: { site: 0, rack: 0, encl: 0, ctrl: 0, disk: 0 }
profiles:
  - name: first-pool
    pools: [ even ]
  - name: both-pools
    pools: [ even, odd ]
```

andriytk commented:
But I/O failed, unfortunately:

```
[vagrant@cmu mcp]$ sudo LD_LIBRARY_PATH=../../../motr/.libs /usr/bin/time ./mcp -prof 0x7000000000000001:0x55 -ep 192.168.180.145@tcp:12345:4:2 -hax 192.168.180.145@tcp:12345:1:1 -proc 0x7200000000000001:0x2b -pool 0x6f00000000000001:0x2e -v /tmp/128M 0x1112243:0x88577801 
motr[05983]:  bed0  ERROR  [pool/pool.c:1812:m0_pools_setup]  <! rc=-2 Cannot find metadata pool <6f00000000000001:4d>
2020/10/22 15:31:17 mio.go:139: m0_client_init() failed: -2
panic: m0_client_init() failed: -2

goroutine 1 [running]:
log.Panicf(0x51afaa, 0x1b, 0xc000040db0, 0x1, 0x1)
	/usr/lib/golang/src/log/log.go:358 +0xc5
_/data/motr/bindings/go/mio.Init()
	/data/motr/bindings/go/mio/mio.go:139 +0x689
main.main()
	/data/motr/bindings/go/mcp/mcp.go:32 +0x5f
0.15user 0.33system 0:01.37elapsed 35%CPU (0avgtext+0avgdata 90276maxresident)k
0inputs+0outputs (0major+20969minor)pagefaults 0swaps
[vagrant@cmu mcp]$ sudo LD_LIBRARY_PATH=../../../motr/.libs /usr/bin/time ./mcp -prof 0x7000000000000001:0x54 -ep 192.168.180.145@tcp:12345:4:2 -hax 192.168.180.145@tcp:12345:1:1 -proc 0x7200000000000001:0x2b -pool 0x6f00000000000001:0x2e -v /tmp/128M 0x1112243:0x88577801 
motr[06803]:  bd00  ERROR  [pool/pool.c:1812:m0_pools_setup]  <! rc=-2 Cannot find metadata pool <6f00000000000001:4d>
2020/10/22 15:32:23 mio.go:139: m0_client_init() failed: -2
panic: m0_client_init() failed: -2

goroutine 1 [running]:
log.Panicf(0x51afaa, 0x1b, 0xc000045db0, 0x1, 0x1)
	/usr/lib/golang/src/log/log.go:358 +0xc5
_/data/motr/bindings/go/mio.Init()
	/data/motr/bindings/go/mio/mio.go:139 +0x689
main.main()
	/data/motr/bindings/go/mcp/mcp.go:32 +0x5f
0.13user 0.24system 0:01.28elapsed 29%CPU (0avgtext+0avgdata 92912maxresident)k
0inputs+0outputs (0major+20961minor)pagefaults 0swaps
```

andriytk commented:
I/O doesn't work even on the single pool configuration:

```
[vagrant@cmu mcp]$ hctl bootstrap --mkfs ~/singlenode.yaml 
2020-10-22 15:36:14: Generating cluster configuration... OK
2020-10-22 15:36:15: Starting Consul server on this node................. OK
2020-10-22 15:36:30: Importing configuration into the KV store... OK
2020-10-22 15:36:31: Starting Consul on other nodes... OK
2020-10-22 15:36:31: Updating Consul configuraton from the KV store... OK
2020-10-22 15:36:32: Waiting for the RC Leader to get elected...... OK
2020-10-22 15:36:35: Starting Motr (phase1, mkfs)... OK
2020-10-22 15:36:42: Starting Motr (phase1, m0d)... OK
2020-10-22 15:36:44: Starting Motr (phase2, mkfs)... OK
2020-10-22 15:36:53: Starting Motr (phase2, m0d)... OK
2020-10-22 15:36:56: Checking health of services... OK
[vagrant@cmu mcp]$ hctl status
Data pool:
    # fid name
    0x6f00000000000001:0x2e 'the pool'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0x4e 'default': 'the pool'
Services:
    localhost  (RC)
    [started]  hax        0x7200000000000001:0x6   192.168.180.145@tcp:12345:1:1
    [started]  confd      0x7200000000000001:0x9   192.168.180.145@tcp:12345:2:1
    [started]  ioservice  0x7200000000000001:0xc   192.168.180.145@tcp:12345:2:2
    [unknown]  m0_client  0x7200000000000001:0x28  192.168.180.145@tcp:12345:4:1
    [unknown]  m0_client  0x7200000000000001:0x2b  192.168.180.145@tcp:12345:4:2
[vagrant@cmu mcp]$ sudo LD_LIBRARY_PATH=../../../motr/.libs /usr/bin/time ./mcp -prof 0x7000000000000001:0x4e -ep 192.168.180.145@tcp:12345:4:1 -hax 192.168.180.145@tcp:12345:1:1 -proc 0x7200000000000001:0x28 -pool 0x6f00000000000001:0x2e -v /tmp/128M 0x1112243:0x88577801
motr[11972]:  8fe0  ERROR  [pool/pool.c:1812:m0_pools_setup]  <! rc=-2 Cannot find metadata pool <6f00000000000001:47>
2020/10/22 15:38:36 mio.go:139: m0_client_init() failed: -2
panic: m0_client_init() failed: -2

goroutine 1 [running]:
log.Panicf(0x51afaa, 0x1b, 0xc000045db0, 0x1, 0x1)
	/usr/lib/golang/src/log/log.go:358 +0xc5
_/data/motr/bindings/go/mio.Init()
	/data/motr/bindings/go/mio/mio.go:139 +0x689
main.main()
	/data/motr/bindings/go/mcp/mcp.go:32 +0x5f
0.13user 0.31system 0:01.33elapsed 33%CPU (0avgtext+0avgdata 92036maxresident)k
0inputs+0outputs (0major+20958minor)pagefaults 0swaps
```

vvv commented:
@andriytk Could you attach `confd.xc` file and the output of `consul kv get -recurse` to this issue please?

andriytk commented:
Sure. Here is the confd-mp.xc for the multi-pool version of singlenode cluster - 
[confd-mp.xc.txt](https://github.com/Seagate/cortx-hare/files/5424226/confd-mp.xc.txt). And here is the same file from the current code which works fine - 
[confd.xc.txt](https://github.com/Seagate/cortx-hare/files/5424235/confd.xc.txt).

Consul kv get:

```
[vagrant@cmu hare]$ consul kv get -recurse m0conf
m0conf/nodes/0x6e00000000000001:0x3:{"name": "localhost", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x28:{"name": "m0_client_other", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x28/services/0x7300000000000001:0x29:{"name": "rms", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x2b:{"name": "m0_client_other", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x2b/services/0x7300000000000001:0x2c:{"name": "rms", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x6:{"name": "hax", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x6/services/0x7300000000000001:0x7:{"name": "ha", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x6/services/0x7300000000000001:0x8:{"name": "rms", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x9:{"name": "m0_server", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x9/services/0x7300000000000001:0xa:{"name": "rms", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x9/services/0x7300000000000001:0xb:{"name": "confd", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc:{"name": "m0_server", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0x23:{"name": "sns_rep", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0x24:{"name": "sns_reb", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0x25:{"name": "addb2", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0x26:{"name": "cas", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0x26/sdevs/0x6400000000000001:0x3f:{"path": "/dev/null", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0x27:{"name": "iscs", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xd:{"name": "rms", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe:{"name": "ios", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0x11:{"path": "/dev/loop1", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0x13:{"path": "/dev/loop2", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0x15:{"path": "/dev/loop3", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0x17:{"path": "/dev/loop4", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0x19:{"path": "/dev/loop5", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0x1b:{"path": "/dev/loop6", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0x1d:{"path": "/dev/loop7", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0x1f:{"path": "/dev/loop8", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0x21:{"path": "/dev/loop9", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0xf:{"path": "/dev/loop0", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/localhost/processes/12/endpoint:192.168.180.145@tcp:12345:2:2
m0conf/nodes/localhost/processes/12/services/addb2:37
m0conf/nodes/localhost/processes/12/services/cas:38
m0conf/nodes/localhost/processes/12/services/ios:14
m0conf/nodes/localhost/processes/12/services/iscs:39
m0conf/nodes/localhost/processes/12/services/rms:13
m0conf/nodes/localhost/processes/12/services/sns_reb:36
m0conf/nodes/localhost/processes/12/services/sns_rep:35
m0conf/nodes/localhost/processes/40/endpoint:192.168.180.145@tcp:12345:4:1
m0conf/nodes/localhost/processes/40/services/m0_client_other:42
m0conf/nodes/localhost/processes/40/services/rms:41
m0conf/nodes/localhost/processes/43/endpoint:192.168.180.145@tcp:12345:4:2
m0conf/nodes/localhost/processes/43/services/m0_client_other:45
m0conf/nodes/localhost/processes/43/services/rms:44
m0conf/nodes/localhost/processes/6/endpoint:192.168.180.145@tcp:12345:1:1
m0conf/nodes/localhost/processes/6/services/ha:7
m0conf/nodes/localhost/processes/6/services/rms:8
m0conf/nodes/localhost/processes/9/endpoint:192.168.180.145@tcp:12345:2:1
m0conf/nodes/localhost/processes/9/services/confd:11
m0conf/nodes/localhost/processes/9/services/rms:10
m0conf/pools/0x6f00000000000001:0x2e:the pool
m0conf/profiles/0x7000000000000001:0x4e:{"name": "default", "pools": ["the pool"]}
m0conf/sites/0x5300000000000001:0x1:{"state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2:{"state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4:{"state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5:{"node": "0x6e00000000000001:0x3", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x10:{"sdev": "0x6400000000000001:0xf", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x12:{"sdev": "0x6400000000000001:0x11", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x14:{"sdev": "0x6400000000000001:0x13", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x16:{"sdev": "0x6400000000000001:0x15", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x18:{"sdev": "0x6400000000000001:0x17", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x1a:{"sdev": "0x6400000000000001:0x19", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x1c:{"sdev": "0x6400000000000001:0x1b", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x1e:{"sdev": "0x6400000000000001:0x1d", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x20:{"sdev": "0x6400000000000001:0x1f", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x22:{"sdev": "0x6400000000000001:0x21", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x40:{"sdev": "0x6400000000000001:0x3f", "state": "M0_NC_UNKNOWN"}
```

andriytk commented:
Consul kv get from the old working cluster (same singlenode CDF, just to compare):

```
[vagrant@cmu hare]$ consul kv get -recurse m0conf
m0conf/nodes/0x6e00000000000001:0x3:{"name": "localhost", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x28:{"name": "m0_client_other", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x28/services/0x7300000000000001:0x29:{"name": "rms", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x2b:{"name": "m0_client_other", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x2b/services/0x7300000000000001:0x2c:{"name": "rms", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x6:{"name": "hax", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x6/services/0x7300000000000001:0x7:{"name": "ha", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x6/services/0x7300000000000001:0x8:{"name": "rms", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x9:{"name": "m0_server", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x9/services/0x7300000000000001:0xa:{"name": "rms", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0x9/services/0x7300000000000001:0xb:{"name": "confd", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc:{"name": "m0_server", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0x23:{"name": "sns_rep", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0x24:{"name": "sns_reb", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0x25:{"name": "addb2", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0x26:{"name": "cas", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0x26/sdevs/0x6400000000000001:0x40:{"path": "/dev/null", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0x27:{"name": "iscs", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xd:{"name": "rms", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe:{"name": "ios", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0x11:{"path": "/dev/loop1", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0x13:{"path": "/dev/loop2", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0x15:{"path": "/dev/loop3", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0x17:{"path": "/dev/loop4", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0x19:{"path": "/dev/loop5", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0x1b:{"path": "/dev/loop6", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0x1d:{"path": "/dev/loop7", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0x1f:{"path": "/dev/loop8", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0x21:{"path": "/dev/loop9", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/0x6e00000000000001:0x3/processes/0x7200000000000001:0xc/services/0x7300000000000001:0xe/sdevs/0x6400000000000001:0xf:{"path": "/dev/loop0", "state": "M0_NC_UNKNOWN"}
m0conf/nodes/localhost/processes/12/endpoint:192.168.180.145@tcp:12345:2:2
m0conf/nodes/localhost/processes/12/services/addb2:37
m0conf/nodes/localhost/processes/12/services/cas:38
m0conf/nodes/localhost/processes/12/services/ios:14
m0conf/nodes/localhost/processes/12/services/iscs:39
m0conf/nodes/localhost/processes/12/services/rms:13
m0conf/nodes/localhost/processes/12/services/sns_reb:36
m0conf/nodes/localhost/processes/12/services/sns_rep:35
m0conf/nodes/localhost/processes/40/endpoint:192.168.180.145@tcp:12345:4:1
m0conf/nodes/localhost/processes/40/services/m0_client_other:42
m0conf/nodes/localhost/processes/40/services/rms:41
m0conf/nodes/localhost/processes/43/endpoint:192.168.180.145@tcp:12345:4:2
m0conf/nodes/localhost/processes/43/services/m0_client_other:45
m0conf/nodes/localhost/processes/43/services/rms:44
m0conf/nodes/localhost/processes/6/endpoint:192.168.180.145@tcp:12345:1:1
m0conf/nodes/localhost/processes/6/services/ha:7
m0conf/nodes/localhost/processes/6/services/rms:8
m0conf/nodes/localhost/processes/9/endpoint:192.168.180.145@tcp:12345:2:1
m0conf/nodes/localhost/processes/9/services/confd:11
m0conf/nodes/localhost/processes/9/services/rms:10
m0conf/profiles:0x7000000000000001:0x2e
m0conf/profiles/pools:0x6f00000000000001:0x2f
m0conf/sites/0x5300000000000001:0x1:{"state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2:{"state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4:{"state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5:{"node": "0x6e00000000000001:0x3", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x10:{"sdev": "0x6400000000000001:0xf", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x12:{"sdev": "0x6400000000000001:0x11", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x14:{"sdev": "0x6400000000000001:0x13", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x16:{"sdev": "0x6400000000000001:0x15", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x18:{"sdev": "0x6400000000000001:0x17", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x1a:{"sdev": "0x6400000000000001:0x19", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x1c:{"sdev": "0x6400000000000001:0x1b", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x1e:{"sdev": "0x6400000000000001:0x1d", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x20:{"sdev": "0x6400000000000001:0x1f", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x22:{"sdev": "0x6400000000000001:0x21", "state": "M0_NC_UNKNOWN"}
m0conf/sites/0x5300000000000001:0x1/racks/0x6100000000000001:0x2/encls/0x6500000000000001:0x4/ctrls/0x6300000000000001:0x5/drives/0x6b00000000000001:0x41:{"sdev": "0x6400000000000001:0x40", "state": "M0_NC_UNKNOWN"}
```

vvv commented:
New `cfgen` doesn't fill 'profile' objects correctly.
```
$ m0confgen --from xcode --to cfgen confd-old.xc | grep '^.prof'
(profile-46 pools=[pool-47, pool-63, pool-72])
$ m0confgen --from xcode --to cfgen confd-new.xc | grep '^.prof'
(profile-78 pools=[pool-46])
```
'profile-78' should refer to three pools, but it only refers to one.

Let me fix that...

vvv commented:
@andriytk I don't have development environment.  Would you test the updated branch please?

vvv commented:
I've rebased the branch onto `main`, removing the merge commit.

I fixed [the problem](https://github.com/Seagate/cortx-hare/pull/1373#issuecomment-714723317), but did not test the result for the lack of \[desire to setup\] the development environment.

<details>
<summary>The fix.</summary>

```diff
diff --git a/cfgen/cfgen b/cfgen/cfgen
index f9938dc..6c5bbf9 100755
--- a/cfgen/cfgen
+++ b/cfgen/cfgen
@@ -1174,8 +1174,10 @@ class ConfPool(ToDhall):
         tolerance = None if t is PoolT.sns else Failures(0, 0, 0, 1, 0)
         pver = ConfPver.build(m0conf, pool_id, pd_attrs, drives, tolerance)
         if t is PoolT.dix:
+            assert m0conf[parent].imeta_pver is None
             m0conf[parent].imeta_pver = pver
         elif t is PoolT.md:
+            assert m0conf[parent].mdpool is None
             m0conf[parent].mdpool = pool_id
         else:
             assert t is PoolT.sns
@@ -1456,10 +1458,22 @@ class ConfProfile(ToDhall):
         pool_ids = dict((m0conf[pool_id]._name, pool_id) for pool_id in m0conf
                         if pool_id.type is ObjT.pool
                         and m0conf[pool_id]._name in prof_desc['pools'])
+        prof_pools = [pool_ids[pool] for pool in prof_desc['pools']]
+
+        root = m0conf[parent]
+        dix_pool = [pool for pool in root.pools
+                    if root.imeta_pver in pool.pvers]
+        assert len(dix_pool) == 1
+        dix_pool = dix_pool[0]
+        if dix_pool not in prof_pools:
+            prof_pools.append(dix_pool)
+
+        assert root.mdpool is not None
+        if root.mdpool not in prof_pools:
+            prof_pools.append(root.mdpool)
+
         prof_id = new_oid(ObjT.profile)
-        m0conf[prof_id] = cls(
-            pools=[pool_ids[pool] for pool in prof_desc['pools']],
-            name=prof_desc['name'])
+        m0conf[prof_id] = cls(pools=prof_pools, name=prof_desc['name'])
         m0conf[parent].profiles.append(prof_id)
         return prof_id
 
```

</details>

andriytk commented:
Thanks @vvv! Will try it today...

andriytk commented:
It does not compile :(

```
[vagrant@cmu hare]$ make
...
--> Checking cfgen
cfgen:1469: error: Argument 1 to "append" of "list" has incompatible type "List[Any]"; expected "Oid"
Found 1 error in 1 file (checked 1 source file)
```

andriytk commented:
Ooops, now another problem:

```
[vagrant@cmu hare]$ hctl bootstrap --mkfs ~/singlenode.yaml 
2020-10-23 11:38:14: Generating cluster configuration...Traceback (most recent call last):
  File "/opt/seagate/cortx/hare/bin/cfgen", line 1741, in <module>
    main()
  File "/opt/seagate/cortx/hare/bin/cfgen", line 139, in main
    cluster = build_cluster(cluster_desc)
  File "/opt/seagate/cortx/hare/bin/cfgen", line 1727, in build_cluster
    ConfProfile.build(conf, root_id, prof)
  File "/opt/seagate/cortx/hare/bin/cfgen", line 1464, in build
    dix_pool = [pool for pool in root.pools
  File "/opt/seagate/cortx/hare/bin/cfgen", line 1465, in <listcomp>
    if root.imeta_pver in pool.pvers]
AttributeError: 'Oid' object has no attribute 'pvers'
```

andriytk commented:
Ok, the i/o works now:

```
[vagrant@cmu hare]$ hctl bootstrap --mkfs ~/singlenode.yaml 
2020-10-23 11:48:41: Generating cluster configuration... OK
2020-10-23 11:48:42: Starting Consul server on this node......... OK
2020-10-23 11:48:50: Importing configuration into the KV store... OK
2020-10-23 11:48:50: Starting Consul on other nodes... OK
2020-10-23 11:48:50: Updating Consul configuraton from the KV store... OK
2020-10-23 11:48:51: Waiting for the RC Leader to get elected......... OK
2020-10-23 11:48:59: Starting Motr (phase1, mkfs)... OK
2020-10-23 11:49:19: Starting Motr (phase1, m0d)... OK
2020-10-23 11:49:22: Starting Motr (phase2, mkfs)... OK
2020-10-23 11:49:38: Starting Motr (phase2, m0d)... OK
2020-10-23 11:49:42: Checking health of services... OK
[vagrant@cmu hare]$ 
[vagrant@cmu hare]$ 
[vagrant@cmu hare]$ hctl status
Data pool:
    # fid name
    0x6f00000000000001:0x2e 'the pool'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0x4e 'default': 'the pool' None None
Services:
    localhost  (RC)
    [started]  hax        0x7200000000000001:0x6   192.168.180.146@tcp:12345:1:1
    [started]  confd      0x7200000000000001:0x9   192.168.180.146@tcp:12345:2:1
    [started]  ioservice  0x7200000000000001:0xc   192.168.180.146@tcp:12345:2:2
    [unknown]  m0_client  0x7200000000000001:0x28  192.168.180.146@tcp:12345:4:1
    [unknown]  m0_client  0x7200000000000001:0x2b  192.168.180.146@tcp:12345:4:2
[vagrant@cmu hare]$ cd /data/motr/bindings/go/mcp/
[vagrant@cmu mcp]$ sudo LD_LIBRARY_PATH=../../../motr/.libs /usr/bin/time ./mcp -prof 0x7000000000000001:0x4e -ep 192.168.180.146@tcp:12345:4:1 -hax 192.168.180.146@tcp:12345:1:1 -proc 0x7200000000000001:0x28 -v /tmp/128M 0x1112243:0x88577801
2020/10/23 11:56:24 mio.go:409: off=0 len=33554432 bs=33554432 gs=4194304 speed=24 (Mbytes/sec)
2020/10/23 11:56:25 mio.go:409: off=33554432 len=33554432 bs=33554432 gs=4194304 speed=22 (Mbytes/sec)
2020/10/23 11:56:27 mio.go:409: off=67108864 len=33554432 bs=33554432 gs=4194304 speed=31 (Mbytes/sec)
2020/10/23 11:56:29 mio.go:409: off=100663296 len=33554432 bs=33554432 gs=4194304 speed=30 (Mbytes/sec)
2.58user 2.61system 0:08.25elapsed 62%CPU (0avgtext+0avgdata 131148maxresident)k
262144inputs+0outputs (0major+24565minor)pagefaults 0swaps
```

andriytk commented:
Multi-pools functionality also works fine:

```
[vagrant@cmu mcp]$ hctl bootstrap --mkfs ~/singlenode-mp.yaml
2020-10-23 12:18:18: Generating cluster configuration... OK
2020-10-23 12:18:19: Starting Consul server on this node.......... OK
2020-10-23 12:18:27: Importing configuration into the KV store... OK
2020-10-23 12:18:27: Starting Consul on other nodes... OK
2020-10-23 12:18:27: Updating Consul configuraton from the KV store... OK
2020-10-23 12:18:28: Waiting for the RC Leader to get elected............ OK
2020-10-23 12:18:38: Starting Motr (phase1, mkfs)... OK
2020-10-23 12:18:44: Starting Motr (phase1, m0d)... OK
2020-10-23 12:18:46: Starting Motr (phase2, mkfs)... OK
2020-10-23 12:18:55: Starting Motr (phase2, m0d)... OK
2020-10-23 12:18:58: Checking health of services... OK
[vagrant@cmu mcp]$ hctl status
Data pools:
    # fid name
    0x6f00000000000001:0x2e 'even'
    0x6f00000000000001:0x3b 'odd'
Profiles:
    # fid name: pool(s)
    0x7000000000000001:0x54 'first-pool': 'even' None None
    0x7000000000000001:0x55 'both-pools': 'even' 'odd' None None
Services:
    localhost  (RC)
    [started]  hax        0x7200000000000001:0x6   192.168.180.146@tcp:12345:1:1
    [started]  confd      0x7200000000000001:0x9   192.168.180.146@tcp:12345:2:1
    [started]  ioservice  0x7200000000000001:0xc   192.168.180.146@tcp:12345:2:2
    [unknown]  m0_client  0x7200000000000001:0x28  192.168.180.146@tcp:12345:4:1
    [unknown]  m0_client  0x7200000000000001:0x2b  192.168.180.146@tcp:12345:4:2
```

Write new object to pool 0x6f00000000000001:0x2e:

```
[vagrant@cmu mcp]$ sudo LD_LIBRARY_PATH=../../../motr/.libs /usr/bin/time ./mcp -prof 0x7000000000000001:0x55 -ep 192.168.180.146@tcp:12345:4:1 -hax 192.168.180.146@tcp:12345:1:1 -proc 0x7200000000000001:0x28 -pool 0x6f00000000000001:0x2e -v /tmp/128M 0x112233:0x456701
2020/10/23 12:21:13 mio.go:409: W: off=0 len=33554432 bs=25165824 gs=12582912 speed=10 (Mbytes/sec)
2020/10/23 12:21:19 mio.go:409: W: off=33554432 len=33554432 bs=25165824 gs=12582912 speed=5 (Mbytes/sec)
2020/10/23 12:21:24 mio.go:409: W: off=67108864 len=33554432 bs=25165824 gs=12582912 speed=5 (Mbytes/sec)
2020/10/23 12:21:27 mio.go:409: W: off=100663296 len=33554432 bs=25165824 gs=12582912 speed=10 (Mbytes/sec)
16.15user 1.07system 0:19.09elapsed 90%CPU (0avgtext+0avgdata 219836maxresident)k
0inputs+0outputs (0major+133403minor)pagefaults 0swaps
```

Write new object to another pool 0x6f00000000000001:0x3b

```
[vagrant@cmu mcp]$ sudo LD_LIBRARY_PATH=../../../motr/.libs /usr/bin/time ./mcp -prof 0x7000000000000001:0x55 -ep 192.168.180.146@tcp:12345:4:1 -hax 192.168.180.146@tcp:12345:1:1 -proc 0x7200000000000001:0x28 -pool 0x6f00000000000001:0x3b -v /tmp/128M 0x112233:0x456702
2020/10/23 12:21:50 mio.go:409: W: off=0 len=33554432 bs=8388608 gs=4194304 speed=26 (Mbytes/sec)
2020/10/23 12:21:51 mio.go:409: W: off=33554432 len=33554432 bs=8388608 gs=4194304 speed=25 (Mbytes/sec)
2020/10/23 12:21:52 mio.go:409: W: off=67108864 len=33554432 bs=8388608 gs=4194304 speed=25 (Mbytes/sec)
2020/10/23 12:21:54 mio.go:409: W: off=100663296 len=33554432 bs=8388608 gs=4194304 speed=24 (Mbytes/sec)
4.34user 0.29system 0:06.42elapsed 72%CPU (0avgtext+0avgdata 127276maxresident)k
0inputs+0outputs (0major+22513minor)pagefaults 0swaps
```

Note the different speed and optimal `bs` and `gs` values are due to different configuration of the pool.

Now let's copy the object to another pool:

```
[vagrant@cmu mcp]$ sudo LD_LIBRARY_PATH=../../../motr/.libs /usr/bin/time ./mcp -prof 0x7000000000000001:0x55 -ep 192.168.180.146@tcp:12345:4:1 -hax 192.168.180.146@tcp:12345:1:1 -proc 0x7200000000000001:0x28 -pool 0x6f00000000000001:0x2e -osz $((128*1024)) -v 0x112233:0x456702 0x112233:0x456704
2020/10/23 12:22:57 mio.go:456: R: off=0 len=33554432 bs=8388608 gs=4194304 speed=54 (Mbytes/sec)
2020/10/23 12:23:00 mio.go:409: W: off=0 len=33554432 bs=25165824 gs=12582912 speed=10 (Mbytes/sec)
2020/10/23 12:23:00 mio.go:456: R: off=33554432 len=33554432 bs=8388608 gs=4194304 speed=56 (Mbytes/sec)
2020/10/23 12:23:06 mio.go:409: W: off=33554432 len=33554432 bs=25165824 gs=12582912 speed=5 (Mbytes/sec)
2020/10/23 12:23:07 mio.go:456: R: off=67108864 len=33554432 bs=8388608 gs=4194304 speed=50 (Mbytes/sec)
2020/10/23 12:23:12 mio.go:409: W: off=67108864 len=33554432 bs=25165824 gs=12582912 speed=5 (Mbytes/sec)
2020/10/23 12:23:13 mio.go:456: R: off=100663296 len=33554432 bs=8388608 gs=4194304 speed=54 (Mbytes/sec)
2020/10/23 12:23:16 mio.go:409: W: off=100663296 len=33554432 bs=25165824 gs=12582912 speed=11 (Mbytes/sec)
18.51user 0.55system 0:20.83elapsed 91%CPU (0avgtext+0avgdata 221200maxresident)k
0inputs+0outputs (0major+133063minor)pagefaults 0swaps
```

CDF:

```
[vagrant@cmu mcp]$ cat ~/singlenode-mp.yaml 
# Cluster Description File (CDF).
# See `cfgen --help-schema` for the format description.

nodes:
  - hostname: localhost     # [user@]hostname
    data_iface: eth0        # name of data network interface
    #data_iface_type: tcp  # type of network interface (optional);
                            # supported values: "tcp" (default), "o2ib"
    m0_servers:
      - runs_confd: true
        io_disks:
          data: []
      - io_disks:
          #meta_data: /path/to/meta-data/drive
          data:
            - /dev/loop0
            - /dev/loop1
            - /dev/loop2
            - /dev/loop3
            - /dev/loop4
            - /dev/loop5
            - /dev/loop6
            - /dev/loop7
            - /dev/loop8
            - /dev/loop9
    m0_clients:
      s3: 0         # number of S3 servers to start
      other: 2      # max quantity of other Motr clients this host may have
pools:
  - name: even
    #type: sns  # optional; supported values: "sns" (default), "dix", "md"
    disk_refs:
      - { path: /dev/loop0, node: localhost }
      - path: /dev/loop1
      - path: /dev/loop2
      - path: /dev/loop3
      - path: /dev/loop4
      - path: /dev/loop6
      - path: /dev/loop8
    data_units: 3
    parity_units: 2
    #allowed_failures: { site: 0, rack: 0, encl: 0, ctrl: 0, disk: 0 }
  - name: odd
    disk_refs:
      - path: /dev/loop5
      - path: /dev/loop7
      - path: /dev/loop9
    data_units: 1
    parity_units: 1
    #allowed_failures: { site: 0, rack: 0, encl: 0, ctrl: 0, disk: 0 }
profiles:
  - name: first-pool
    pools: [ even ]
  - name: both-pools
    pools: [ even, odd ]
```
